src/main/java/javaslang/collection/Vector.java

```diff
@@ -0,0 +1,1481 @@
+
+        @Override
+        public boolean equals(Object o) {
+            if (o == this) {
+                return true;
+            } else if (o instanceof Vector) {
+                Vector<?> vector1 = this;
+                Vector<?> vector2 = (Vector<?>) o;
+                while (!vector1.isEmpty() && !vector2.isEmpty()) {
+                    final boolean isEqual = Objects.equals(vector1.head(), vector2.head());
+                    if (!isEqual) {
+                        return false;
+                    }
+                    vector1 = vector1.tail();
+                    vector2 = vector2.tail();
+                }
+                return vector1.isEmpty() && vector2.isEmpty();
+            } else {
+                return false;
+            }
+        }
+
+        @Override
+        public int hashCode() {
+            return hashCode.get();
+        }
+
+        @Override
+        public String toString() {
+            return map(String::valueOf).join(", ", "Vector(", ")");
+        }
+
+        @Override
+        public HashArrayMappedTrie<Integer, T> toHashArrayMappedTrie() {
+            return trie;
+        }
+
+        private Object writeReplace() {
+            return new SerializationProxy<>(this);
+        }
+
+        private void readObject(ObjectInputStream stream) throws InvalidObjectException {
+            throw new InvalidObjectException("Proxy required");
+        }
+
+        /**
+         * A serialization proxy.
+         * See {@link List.Cons.SerializationProxy}
+         */
+        private static final class SerializationProxy<T> implements Serializable {
+
+            private static final long serialVersionUID = 1L;
+
+            private transient Impl<T> vector;
+
+            SerializationProxy(Impl<T> vector) {
+                this.vector = vector;
+            }
+
+            private void writeObject(ObjectOutputStream s) throws IOException {
+                s.defaultWriteObject();
+                s.writeInt(vector.length());
+                for (T t: vector) {
+                    s.writeObject(t);
+                }
+            }
+
+            private void readObject(ObjectInputStream s) throws ClassNotFoundException, IOException {
+                s.defaultReadObject();
+                final int size = s.readInt();
+                if (size <= 0) {
+                    throw new InvalidObjectException("No elements");
+                }
+                HashArrayMappedTrie<Integer, T> temp = HashArrayMappedTrie.empty();
+                for (int i = 0; i < size; i++) {
+                    @SuppressWarnings("unchecked")
+                    final T element = (T) s.readObject();
+                    temp = temp.put(temp.size(), element);
+                }
+                vector = new Impl<>(temp);
+            }
+
+            private Object readResolve() {
+                return vector;
+            }
+        }
+    }
+
+    class Nil<T> implements Vector<T>, Serializable {
```

Comment:
And _boom_, Ruslan comes with Vector around the corner :-)) I love this HAMT!!

I've learned that Nil is somewhat 'reserved' for the empty (lazy) linked list. Lets name it `Empty`.